From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Tue, 14 Jan 2020 14:59:08 -0800
Subject: [PATCH] Optimise Chunk#getFluid

Removing the try catch and generally reducing ops should make it
faster on its own, however removing the try catch makes it
easier to inline due to code size

diff --git a/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java b/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
index 1a64618e533426147ad6ab1211afb9cc676cc185..f994d393929ca6813d4209fb9c93550e4f692228 100644
--- a/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/src/main/java/net/minecraft/world/level/chunk/LevelChunk.java
@@ -373,18 +373,20 @@ public class LevelChunk extends ChunkAccess {
     }
 
     public FluidState getFluidState(int x, int y, int z) {
-        try {
-            int l = this.getSectionIndex(y);
-
-            if (l >= 0 && l < this.sections.length) {
-                LevelChunkSection chunksection = this.sections[l];
+        // try {  // Paper - remove try catch
+        // Paper start - reduce the number of ops in this call
+        int index = this.getSectionIndex(y);
+            if (index >= 0 && index < this.sections.length) {
+                LevelChunkSection chunksection = this.sections[index];
 
                 if (!chunksection.hasOnlyAir()) {
-                    return chunksection.getFluidState(x & 15, y & 15, z & 15);
+                    return chunksection.states.get((y & 15) << 8 | (z & 15) << 4 | x & 15).getFluidState();
+                    // Paper end
                 }
             }
 
             return Fluids.EMPTY.defaultFluidState();
+        /* // Paper - remove try catch
         } catch (Throwable throwable) {
             CrashReport crashreport = CrashReport.forThrowable(throwable, "Getting fluid state");
             CrashReportCategory crashreportsystemdetails = crashreport.addCategory("Block being got");
@@ -394,6 +396,7 @@ public class LevelChunk extends ChunkAccess {
             });
             throw new ReportedException(crashreport);
         }
+         */  // Paper - remove try catch
     }
 
     // CraftBukkit start
diff --git a/src/main/java/net/minecraft/world/level/chunk/LevelChunkSection.java b/src/main/java/net/minecraft/world/level/chunk/LevelChunkSection.java
index 67b06640358e04752bf149e9380f5f83d5d1d0f6..cdd17e501c678a4f4bebbaaccdaec1682351e2f2 100644
--- a/src/main/java/net/minecraft/world/level/chunk/LevelChunkSection.java
+++ b/src/main/java/net/minecraft/world/level/chunk/LevelChunkSection.java
@@ -52,7 +52,7 @@ public class LevelChunkSection {
     }
 
     public FluidState getFluidState(int x, int y, int z) {
-        return ((BlockState) this.states.get(x, y, z)).getFluidState();
+        return this.states.get(x, y, z).getFluidState(); // Paper - diff on change - we expect this to be effectively just getType(x, y, z).getFluid(). If this changes we need to check other patches that use IBlockData#getFluid.
     }
 
     public void acquire() {
